boot/bootutil/src/loader.c

```diff
@@ -1607,158 +1998,178 @@ boot_swap_if_needed(int *out_swap_type)
 int
 boot_go(struct boot_rsp *rsp)
 {
-    int swap_type;
+    size_t img_cnt;
     size_t slot;
+    struct boot_status bs;
     int rc;
     int fa_id;
-    bool reload_headers = false;
 
     /* The array of slot sectors are defined here (as opposed to file scope) so
      * that they don't get allocated for non-boot-loader apps.  This is
      * necessary because the gcc option "-fdata-sections" doesn't seem to have
      * any effect in older gcc versions (e.g., 4.8.4).
      */
-    static boot_sector_t slot0_sectors[BOOT_MAX_IMG_SECTORS];
-    static boot_sector_t slot1_sectors[BOOT_MAX_IMG_SECTORS];
+    static boot_sector_t
+        primary_slot_sectors[BOOT_IMAGE_NUMBER][BOOT_MAX_IMG_SECTORS];
+    static boot_sector_t
+        secondary_slot_sectors[BOOT_IMAGE_NUMBER][BOOT_MAX_IMG_SECTORS];
     static boot_sector_t scratch_sectors[BOOT_MAX_IMG_SECTORS];
-    boot_data.imgs[0].sectors = slot0_sectors;
-    boot_data.imgs[1].sectors = slot1_sectors;
-    boot_data.scratch.sectors = scratch_sectors;
+
+    /* Iterate over all the images. By the end of the loop the swap type has
+     * to be determined for each image and all aborted swaps have to be
+     * completed.
+     */
+    for (img_cnt = 0; img_cnt < BOOT_IMAGE_NUMBER; ++img_cnt) {
+        boot_set_active_image(img_cnt);
+
+        boot_data->imgs[BOOT_PRIMARY_SLOT].sectors =
+                                            primary_slot_sectors[img_cnt];
+        boot_data->imgs[BOOT_SECONDARY_SLOT].sectors =
+                                            secondary_slot_sectors[img_cnt];
+
+        boot_data->scratch.sectors = scratch_sectors;
 
 #ifdef MCUBOOT_ENC_IMAGES
-    /* FIXME: remove this after RAM is cleared by sim */
-    boot_enc_zeroize();
+        /* FIXME: remove this after RAM is cleared by sim */
+        boot_enc_zeroize();
```

Comment:
Does this need to be done here now, since it's also done in `boot_set_active_image()`?

Reply:
No, I've removed it.